bokehjs/src/lib/core/properties.ts

```diff
@@ -447,23 +448,26 @@ export abstract class UnitsSpec<T, Units> extends VectorSpec<T, Dimensional<Vect
   abstract get default_units(): Units
   abstract get valid_units(): Units[]
 
+  spec: Spec & {units?: Units}
+
   _update(attr_value: any): void {
     super._update(attr_value)
 
-    if (this.spec.units == null)
-      this.spec.units = this.default_units
-
-    const units = this.spec.units
-    if (!includes(this.valid_units, units))
+    const {units} = this.spec
+    if (units != null && !includes(this.valid_units, units)) {
       throw new Error(`units must be one of ${this.valid_units.join(", ")}; got: ${units}`)
+    }
   }
 
   get units(): Units {
-    return this.spec.units as Units
+    return this.spec.units ?? this.default_units
   }
 
   set units(units: Units) {
-    this.spec.units = units
+    if (units != this.default_units)
+      this.spec.units = units
+    else
+      delete this.spec.units
```

Comment:
Why this complicated implicit relationship between get/set for default?

Reply:
To make sure we don't send default units around unnecessarily. In future I will drop `spec` in favour of simple class members, so this complexity will go away.

Reply:
right I guess I am wondering if maybe it's just better to send the units all the time. In any case, if you mean narrowly "drop the `.spec` attribute"  (think so?) then 👍 if you are talking about "DataSpec" in general that needs planning/discussion

Reply:
At this point I meant `Property` (or `PrimitiveProperty`), because right now the design exposes details about subclasses, that primitive properties should not know about, among other things (e.g. weak typing).